### PR TITLE
Rustdoc-Json: Add enum discriminant

### PR DESCRIPTION
**Pull request by aDotInTheVoid**
Does the first part of #101337, by adding it to `clean`, but doesn't change HTML output, as 

1. [No Consensus has appeared on the UI for this](https://rust-lang.zulipchat.com/#narrow/stream/266220-rustdoc/topic/Enum.20discriminant.20values.20in.20HTML.20output)
2. [When inlining across crates, information is lost](https://rust-lang.zulipchat.com/#narrow/stream/266220-rustdoc/topic/.60clean_variant_def.20.60vs.20.60clean_variant_data.60)

JSON doesn't have either of these limitations.

r? @GuillaumeGomez

### CODE REVIEW

File: src/rustdoc-json-types/lib.rs [hunk 1]
```diff
@@ -308,11 +308,28 @@ pub struct Enum {
...
+#[derive(Clone, Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
+pub struct Discriminant {
+    /// The expression that produced the discriminant.
+    ///
+    /// Unline `value`, this preserves the origional formatting (eg suffixes,
+    /// hexidecimal, and underscors), making it unsuitable to be machine
+    /// interprited.
+    ///
+    /// In some cases, when the value is to complex, this may be `"{ _ }"`.
+    /// When this occours is unstable, and may change without notice.
```
Comment (fbstj) [inline src/rustdoc-json-types/lib.rs:325]:
```suggestion
    /// When this occurs it is unstable, and may change without notice.
```

File: src/rustdoc-json-types/lib.rs [hunk 2]
```diff
@@ -308,11 +308,28 @@ pub struct Enum {
...
     Struct(Vec<Id>),
 }
 
+#[derive(Clone, Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
+pub struct Discriminant {
+    /// The expression that produced the discriminant.
+    ///
+    /// Unline `value`, this preserves the origional formatting (eg suffixes,
+    /// hexidecimal, and underscors), making it unsuitable to be machine
+    /// interprited.
```
Comment (fbstj) [inline src/rustdoc-json-types/lib.rs:322]:
```suggestion
    /// Unlike `value`, this preserves the original formatting (eg suffixes,
    /// hexadecimal, and underscores), making it unsuitable to be machine
    /// interpreted.
```

File: src/rustdoc-json-types/lib.rs [hunk 3]
```diff
@@ -308,11 +308,28 @@ pub struct Enum {
...
+    ///
+    /// In some cases, when the value is to complex, this may be `"{ _ }"`.
+    /// When this occours is unstable, and may change without notice.
+    pub expr: String,
+    /// The numerical value of the discriminant. Stored as a string due to
+    /// JSON's poor suport for large integers, and the fact that it would need
```
Comment (fbstj) [inline src/rustdoc-json-types/lib.rs:328]:
```suggestion
    /// JSON's poor support for large integers, and the fact that it would need
```

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
Looks good to me. Let's make a perf run first.

@bors try @rust-timer queue

**Comment by GuillaumeGomez** [comment]
It's a bit more than I expected. Considering it's only used for the JSON output, it seems to have a big impact on HTML output. Would it be possible to query this information on demand instead of storing it?

**Comment by GuillaumeGomez** [comment]
Seems all good now. The impact on rss is expected unfortunately, but not much we can do. Thanks for working on this!

@bors r+ rollup=iffy

**Pull request merged**